src/DotNetWorker.Core/Context/Features/DefaultFunctionInputBindingFeature.cs

```diff
@@ -0,0 +1,131 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Azure.Functions.Worker.Converters;
+using Microsoft.Extensions.DependencyInjection;
+
+namespace Microsoft.Azure.Functions.Worker.Context.Features
+{
+    internal class DefaultFunctionInputBindingFeature : IFunctionInputBindingFeature, IDisposable
+    {
+        private const int WaitTimeInMilliSeconds = 100;
+        private readonly SemaphoreSlim _semaphoreSlim = new(1, 1);
+        private readonly IConverterContextFactory _converterContextFactory;
+        private FunctionInputBindingResult? _inputBindingResult;
+        private bool _disposed;
+
+        public DefaultFunctionInputBindingFeature(IConverterContextFactory converterContextFactory)
+        {
+            _converterContextFactory = converterContextFactory ??
+                                       throw new ArgumentNullException(nameof(converterContextFactory));
+        }
+
+        public async ValueTask<FunctionInputBindingResult> BindFunctionInputAsync(FunctionContext context)
+        {
+            ObjectDisposedThrowHelper.ThrowIf(_disposed, this);
+
+            await _semaphoreSlim.WaitAsync(WaitTimeInMilliSeconds, context.CancellationToken);
+
+            if (_inputBindingResult is not null)
+            {
+                // Return the cached value if BindFunctionInputAsync is called a second time during invocation.
+                return _inputBindingResult!;
```

Comment:
if we get here... is the semaphore ever released? should all of this be inside the `try/finally`?

Reply:
Oh yes, Missed that :( Moved now to try block in the latest iteration.

This method grew a bit. Probably can comeback to it later to refactor to smaller methods.